src/UseCases/CreatePayment/CreatePaymentUseCase.php

```diff
@@ -0,0 +1,133 @@
+<?php
+declare( strict_types=1 );
+
+namespace WMDE\Fundraising\PaymentContext\UseCases\CreatePayment;
+
+use WMDE\Euro\Euro;
+use WMDE\Fundraising\PaymentContext\Domain\Model\CreditCardPayment;
+use WMDE\Fundraising\PaymentContext\Domain\Model\Payment;
+use WMDE\Fundraising\PaymentContext\Domain\Model\PaymentInterval;
+use WMDE\Fundraising\PaymentContext\Domain\Model\PayPalPayment;
+use WMDE\Fundraising\PaymentContext\Domain\Model\SofortPayment;
+use WMDE\Fundraising\PaymentContext\Domain\PaymentRepository;
+use WMDE\Fundraising\PaymentContext\Domain\Repositories\PaymentIDRepository;
+use WMDE\Fundraising\PaymentContext\Domain\TransferCodeGenerator;
+
+class CreatePaymentUseCase {
+	public function __construct(
+		private PaymentIDRepository $idGenerator,
+		private PaymentRepository $paymentRepository,
+		private TransferCodeGenerator $paymentReferenceCodeGenerator ) {
+	}
+
+	public function createPayment( PaymentCreationRequest $request ): SuccessResponse|FailureResponse {
+		$paymentOrFailure = $this->doCreatePayment( $request );
+
+		if ( $paymentOrFailure instanceof FailureResponse ) {
+			return $paymentOrFailure;
+		}
+		$payment = $paymentOrFailure;
+
+		$this->paymentRepository->storePayment( $payment );
+		return new SuccessResponse( $this->getNextIdOnce() );
+	}
+
+	private function getNextIdOnce(): int {
+		static $id = null;
+		if ( $id === null ) {
+			$id = $this->idGenerator->getNewID();
+		}
+		return $id;
+	}
+
+	private function doCreatePayment( PaymentCreationRequest $request ): Payment|FailureResponse {
+		try {
+			return match ( $request->paymentType ) {
+				'MCP' => $this->createCreditCardPayment( $request ),
+				'PPL' => $this->createPayPalPayment( $request ),
+				'SUB' => $this->createSofortPayment( $request ),
+				default => new FailureResponse( "Invalid payment type: " . $request->paymentType ),
```

Comment:
Creating the `FailureResponse` here is outside the scope of what that object is supposed to be IMO. This method should throw an exception for the parent caller to catch which then instantiates and returns the failure.

Reply:
Thank you! Your suggestion really made the code better. I have written a detailed explanation of the change in the PR message: https://github.com/wmde/fundraising-payments/pull/72/commits/46b65994fea27f92d0e702188292db4ed618d58a 

I still have this instinct/habit of trying to avoid exceptions for anything that is not truly exceptional. Especially since I have seen the "Result" data type in Rust or the result monad in functional programming. But in this case it makes the code so much nicer and consistent.